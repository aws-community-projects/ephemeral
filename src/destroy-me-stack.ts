import { StackProps, Duration, Stack } from 'aws-cdk-lib';
import { DestroyMeConstruct } from './destroy-me';
import { Construct } from 'constructs';

export interface DestroyMeStackProps extends StackProps {
  readonly destroyMeDuration?: Duration
  readonly destroyMeEnable?: boolean
}
export class DestroyMeStack extends Stack {
  constructor (
    scope: Construct,
    id: string,
    props: DestroyMeStackProps = {
    } as DestroyMeStackProps,
  ) {
    super(scope, id, props);
    const {
      destroyMeEnable: enable,
      destroyMeDuration: duration,
    } = props;

    if (enable) {
      new DestroyMeConstruct(this, 'DestroyMeConstruct', {
        duration: duration ?? Duration.days(7),
      });
    } else {
      console.log('DestroyMe: disabled');
    }
  }
}
