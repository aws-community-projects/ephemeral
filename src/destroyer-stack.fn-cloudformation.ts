import type { EventBridgeEvent } from 'aws-lambda';
import { CloudFormationClient,
  DescribeStacksCommand } from '@aws-sdk/client-cloudformation';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb'; // ES6 import
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';

const marshallOptions = {
  // Whether to automatically convert empty strings, blobs, and sets to `null`.
  convertEmptyValues: false, // false, by default.
  // Whether to remove undefined values while marshalling.
  removeUndefinedValues: true, // false, by default.
  // Whether to convert typeof object to map attribute.
  convertClassInstanceToMap: true, // false, by default.
};

const unmarshallOptions = {
  // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
  wrapNumbers: false, // false, by default.
};

const translateConfig = {
  marshallOptions,
  unmarshallOptions,
};

const client = new DynamoDBClient({
});
const ddbDocClient = DynamoDBDocumentClient.from(client, translateConfig);
const cf = new CloudFormationClient({
});

export const handler = async (
  event: EventBridgeEvent<string, any>,
): Promise<void> => {
  if (!process.env.DESTROY_TABLE_NAME) {
    return;
  }
  const StackName = event.detail['stack-id'];
  const describeCommand = new DescribeStacksCommand({
    StackName,
  });
  const stacks = await cf.send(describeCommand);

  const stack = stacks.Stacks?.[0];

  const stackLife = stack?.Tags?.find((tag) => tag.Key === 'STACK_LIFE')?.Value;
  if (stackLife) {
    try {
      await ddbDocClient.send(
        new PutCommand({
          TableName: process.env.DESTROY_TABLE_NAME,
          Item: {
            pk: stack.StackName,
            ttl: Math.ceil(new Date().getTime() / 1000 + Number(stackLife)),
          },
        }),
      );
    } catch (e) {
      console.log(e);
    }
  }
};
