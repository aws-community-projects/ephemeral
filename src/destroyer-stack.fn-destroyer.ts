import type { DynamoDBStreamEvent } from 'aws-lambda';
import { unmarshall } from '@aws-sdk/util-dynamodb';
import { AttributeValue } from '@aws-sdk/client-dynamodb';
import { CloudFormationClient, DeleteStackCommand } from '@aws-sdk/client-cloudformation';

export const handler = async (event: DynamoDBStreamEvent): Promise<void> => {
  const stacksToDestroy = event.Records.reduce<string[]>((p, record) => {
    if (record.eventName === 'REMOVE') {
      if (record.dynamodb?.OldImage) {
        const oldImage = record.dynamodb.OldImage as Record<
        string,
        AttributeValue
        >;
        const item = unmarshall(oldImage);
        const currentTimeInSeconds = new Date().getTime() / 1000;
        if (item.ttl > currentTimeInSeconds) {
          // item was manually removed and not expired
          console.log('item was manually removed and not expired', currentTimeInSeconds, item.ttl);
          return [...p];
        }
        return [...p, item.pk];
      }
    }
    return [...p];
  }, []);
  console.log(stacksToDestroy);
  if (stacksToDestroy.length > 0) {
    const client = new CloudFormationClient({
    });
    await Promise.all(
      stacksToDestroy.map(
        async (stackName) =>
          await client.send(
            new DeleteStackCommand({
              StackName: stackName,
            }),
          ),
      ),
    );
  }
};
