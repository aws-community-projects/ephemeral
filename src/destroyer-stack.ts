import { Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType,
  BillingMode,
  StreamViewType,
  Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { Runtime, StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { PolicyStatement, Effect } from 'aws-cdk-lib/aws-iam';
import { Rule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction, SnsTopic } from 'aws-cdk-lib/aws-events-targets';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { existsSync } from 'fs';
import { join } from 'path';

export class DestroyerStack extends Stack {
  constructor (scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const tableName = 'destroyer';
    const table = new Table(this, tableName, {
      tableName,
      partitionKey: {
        name: 'pk',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
      timeToLiveAttribute: 'ttl',
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const local = existsSync(join(__dirname, 'destroyer-stack.fn-cloudformation.ts'));
    const cloudformationFn = new NodejsFunction(this, 'fn-cloudformation', {
      runtime: Runtime.NODEJS_18_X,
      memorySize: 1024,
      timeout: Duration.minutes(5),
      entry: join(__dirname, local ? 'destroyer-stack.fn-cloudformation.ts' : 'destroyer-stack.fn-cloudformation.js'),
      initialPolicy: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'cloudformation:Describe*',
            'cloudformation:Get*',
            'cloudformation:List*',
          ],
          resources: ['*'],
        }),
      ],
    });
    table.grantReadWriteData(cloudformationFn);
    cloudformationFn.addEnvironment('DESTROY_TABLE_NAME', table.tableName);
    new Rule(this, 'cloudformation-rule', {
      eventPattern: {
        source: ['aws.cloudformation'],
        detailType: ['CloudFormation Stack Status Change'],
      },
      targets: [new LambdaFunction(cloudformationFn)],
    });

    const destroyFn = new NodejsFunction(this, 'fn-destroyer', {
      runtime: Runtime.NODEJS_18_X,
      memorySize: 1024,
      timeout: Duration.minutes(5),
      entry: join(__dirname, local ? 'destroyer-stack.fn-destroyer.ts' : 'destroyer-stack.fn-destroyer.js'),
    });

    destroyFn.addEventSource(
      new DynamoEventSource(table, {
        startingPosition: StartingPosition.LATEST,
      }),
    );
    destroyFn.addToRolePolicy(
      new PolicyStatement({
        actions: ['cloudformation:DeleteStack'],
        resources: ['*'],
        effect: Effect.ALLOW,
      }),
    );

    const failTopic = new Topic(this, 'fail-topic');
    new Rule(this, 'delete-failed-rule', {
      eventPattern: {
        source: ['aws.cloudformation'],
        detailType: ['CloudFormation Stack Status Change'],
        detail: {
          resourceStatus: ['DELETE_FAILED'],
        },
      },
      targets: [new SnsTopic(failTopic)],
    });
  }
}
