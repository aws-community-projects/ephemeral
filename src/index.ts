export * from './self-destruct-aspect';
export * from './self-destruct-construct';
export * from './self-destruct-stack';
export * from './make-destroyable-aspect';
export * from './destroy-me-construct';
export * from './destroy-me-stack';
export * from './destroyer-stack';
