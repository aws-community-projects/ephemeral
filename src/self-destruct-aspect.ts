import { MakeDestroyableAspect as SelfDestructAspect } from './make-destroyable-aspect';
// backwards compatible
export { SelfDestructAspect };
