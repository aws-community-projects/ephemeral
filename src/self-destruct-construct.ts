import { CallAwsService } from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Choice,
  Condition,
  Map,
  Pass,
  StateMachine,
  Succeed,
  Wait,
  WaitTime } from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';
import { Duration, Stack } from 'aws-cdk-lib';
import { AwsCustomResource,
  AwsCustomResourcePolicy,
  PhysicalResourceId } from 'aws-cdk-lib/custom-resources';

export interface SelfDestructConstructProps {
  readonly duration: Duration
}

export class SelfDestructConstruct extends Construct {
  constructor (scope: Construct, id: string, props: SelfDestructConstructProps = {
  } as SelfDestructConstructProps) {
    super(scope, id);

    const { duration } = props;

    const listExecutions = new CallAwsService(this, 'ListExecutions', {
      action: 'listExecutions',
      iamAction: 'states:ListExecutions',
      iamResources: ['*'],
      parameters: {
        'StateMachineArn.$': '$$.StateMachine.Id',
        StatusFilter: 'RUNNING',
      },
      service: 'sfn',
    });

    const executionsMap = new Map(this, 'ExecutionsMap', {
      inputPath: '$.Executions',
    });

    const stopExecution = new CallAwsService(this, 'StopExecution', {
      action: 'stopExecution',
      iamAction: 'states:StopExecution',
      iamResources: ['*'],
      parameters: {
        Cause: 'Superceded',
        'ExecutionArn.$': '$.ExecutionArn',
      },
      service: 'sfn',
    });

    executionsMap.iterator(
      new Choice(this, 'NotSelf?')
        .when(
          Condition.not(
            Condition.stringEqualsJsonPath('$.ExecutionArn', '$$.Execution.Id'),
          ),
          stopExecution,
        )
        .otherwise(new Pass(this, 'self')),
    );

    const wait = new Wait(this, 'Wait', {
      time: WaitTime.duration(duration),
    });
    const wasDelete = new Choice(this, 'WasDelete?')
      .when(
        Condition.stringEquals('$$.Execution.Input.Action', 'Delete'),
        new Succeed(this, 'DeleteSuccess'),
      )
      .otherwise(wait);

    const deleteStack = new CallAwsService(this, 'DeleteStack', {
      action: 'deleteStack',
      iamAction: 'cloudformation:DeleteStack',
      iamResources: ['*'],
      parameters: {
        'StackName.$': '$$.Execution.Input.StackName',
      },
      service: 'cloudformation',
    });

    const finished = new Succeed(this, 'Finished');

    listExecutions.next(executionsMap);
    executionsMap.next(wasDelete);
    wait.next(deleteStack);
    deleteStack.next(finished);

    const sm = new StateMachine(this, 'SelfDestructMachine', {
      definition: listExecutions,
    });

    new AwsCustomResource(this, 'SelfDestructCR', {
      onCreate: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Create',
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      onDelete: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Delete',
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      onUpdate: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Update',
            Version: new Date().getTime().toString(),
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: [sm.stateMachineArn],
      }),
    });
  }
}
