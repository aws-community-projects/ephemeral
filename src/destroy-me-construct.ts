import { Aspects, Duration, Stack, Tags } from 'aws-cdk-lib';
import { SelfDestructAspect } from './self-destruct-aspect';
import { Construct } from 'constructs';

export interface DestroyMeConstructProps {
  readonly duration: Duration
}

export class DestroyMeConstruct extends Construct {
  constructor (scope: Construct, id: string, props: DestroyMeConstructProps = {
  } as DestroyMeConstructProps) {
    super(scope, id);

    const { duration } = props;
    Tags.of(Stack.of(this)).add('STACK_LIFE', duration.toSeconds().toString());

    Aspects.of(Stack.of(this)).add(new SelfDestructAspect());
  }
}
