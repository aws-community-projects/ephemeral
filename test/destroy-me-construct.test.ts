import { App, Duration, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { DestroyMeConstruct } from '../src';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';

describe('DestroyMeConstruct', () => {
  test('Stack Tagged + Aspect', () => {
    const app = new App();
    const stack = new Stack(app, 'TestStack');
    new Function(stack, 'MyTestFunction', {
      handler: 'index.handler',
      code: Code.fromInline('exports.handler = async () => { return { statusCode: 200, body: "Hello World" }; };'),
      runtime: Runtime.NODEJS_18_X,
    });
    new DestroyMeConstruct(stack, 'MyTestConstruct', {
      duration: Duration.days(1),
    });
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
  });
});
