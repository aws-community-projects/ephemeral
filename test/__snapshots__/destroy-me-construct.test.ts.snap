// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestroyMeConstruct Stack Tagged + Aspect 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyTestFunctionE5665775": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "MyTestFunctionServiceRole1183FD15",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "exports.handler = async () => { return { statusCode: 200, body: \\"Hello World\\" }; };",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestFunctionServiceRole1183FD15",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": Array [
          Object {
            "Key": "STACK_LIFE",
            "Value": "86400",
          },
        ],
      },
      "Type": "AWS::Lambda::Function",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestFunctionServiceRole1183FD15": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "STACK_LIFE",
            "Value": "86400",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
